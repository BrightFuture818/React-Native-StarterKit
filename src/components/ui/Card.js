/**
 * Buttons
 *
     <Button text={'Server is down'} />
 *
 * React Native Starter App
 * https://github.com/mcnamee/react-native-starter-app
 */
import React, { Component } from 'react';
import { Card } from 'react-native-elements';

// Consts and Libs
import { AppSizes } from '@theme/';

/* Component ==================================================================== */
class CustomCard extends Component {
  cardProps = () => {
    // Defaults
    const props = {
      containerStyle: {
        borderRadius: AppSizes.borderRadius,
      },
      ...this.props,
    };

    return props;
  }

  render = () => (
    <Card {...this.cardProps()} />
  )
}

/* Export Component ==================================================================== */
export default CustomCard;
