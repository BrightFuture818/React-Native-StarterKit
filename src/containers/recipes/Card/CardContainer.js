/**
 * Individual Recipe Card Container
 *
 * React Native Starter App
 * https://github.com/mcnamee/react-native-starter-app
 */
import React, { Component, PropTypes } from 'react';
import { connect } from 'react-redux';
import { Actions } from 'react-native-router-flux';

// Actions
import * as UserActions from '@redux/user/actions';

// Components
import RecipeCardRender from './CardView';

/* Redux ==================================================================== */
// What data from the store shall we send to the component?
const mapStateToProps = state => ({
  user: state.user,
});

// Any actions to map to the component?
const mapDispatchToProps = {
  updateFavourites: UserActions.updateMe,
};

/* Component ==================================================================== */
class RecipeCard extends Component {
  static componentName = 'RecipeCard';

  static propTypes = {
    recipe: PropTypes.shape({
      id: PropTypes.number.isRequired,
      title: PropTypes.string.isRequired,
      body: PropTypes.string.isRequired,
      image: PropTypes.string,
    }).isRequired,
    updateFavourites: PropTypes.func.isRequired,
    user: PropTypes.shape({
      acf: PropTypes.shape({
        favourite_recipes: PropTypes.arrayOf(PropTypes.object),
      }),
    }),
  }

  static defaultProps = {
    user: null,
  }

  constructor(props) {
    super(props);
    this.state = { recipe: props.recipe };
  }

  componentWillReceiveProps(props) {
    if (props.recipe) {
      this.setState({ recipe: props.recipe });
    }
  }

  /**
    * On Press of Card
    */
  onPressCard = () => {
    Actions.recipeView({
      title: this.props.recipe.title,
      recipe: this.props.recipe,
    });
  }

  /**
    * When user taps to favourite a recipe
    */
  onPressFavourite = () => {
    const recipeId = this.props.recipe.id;

    if (recipeId && this.props.updateFavourites) {
      const favs = (
        this.props.user && this.props.user.acf &&
        this.props.user.acf.favourite_recipes
      )
        ? this.props.user.acf.favourite_recipes
        : null;

      // Build Payload - Update current user favourites
      // Payload should be:
      //   {fields: {favourite_recipes: [{recipe_id: 43}, {recipe_id: 44}]}}
      const arrIdx = this.isFavourite();

      // Remove from current list
      if (arrIdx) favs.splice(arrIdx, 1);
      // Add to current list
      else favs.push({ recipe_id: recipeId });

      // Send new list to API
      this.props.updateFavourites({ fields: { favourite_recipes: favs } });
    }
  }


  /**
    * Check in Redux to find if this Recipe ID is a Favourite
    */
  isFavourite = () => {
    const { user, recipe } = this.props;

    if (recipe && recipe.id) {
      if (user && user.acf && user.acf.favourite_recipes) {
        const recipeId = recipe.id;
        const favs = user.acf.favourite_recipes;

        // Return true if this recipe is in favourites list
        let result = false;
        favs.forEach((obj, key) => {
          if (String(recipeId) === String(obj.recipe_id)) {
            result = key;
          }
        });

        return result;
      }
    }

    return false;
  }

  render = () => {
    const { recipe } = this.state;
    const { user } = this.props;

    return (
      <RecipeCardRender
        title={recipe.title}
        body={recipe.body}
        image={recipe.image}
        onPress={this.onPressCard}
        onPressFavourite={(user && user.id) ? this.onPressFavourite : null}
        isFavourite={(user && user.id && this.isFavourite()) && true}
      />
    );
  }
}

/* Export Component ==================================================================== */
export default connect(mapStateToProps, mapDispatchToProps)(RecipeCard);
