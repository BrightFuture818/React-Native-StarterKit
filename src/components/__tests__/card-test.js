/**
 * Test to check if the component renderes correctly
 */
/* global it expect */

import 'react-native';
import React from 'react';
import renderer from 'react-test-renderer';

import Card from '@ui/Card';

/**
 * Check if card renders correcly
 * and asserting it to the matching snapshot
 */
it('Card renders correcly', () => {
  const tree = renderer.create(
    <Card />,
  ).toJSON();

  expect(tree).toMatchSnapshot();
});
