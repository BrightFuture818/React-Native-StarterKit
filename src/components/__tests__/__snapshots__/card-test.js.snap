exports[`test Card renders correcly 1`] = `
<View
  style={
    Array [
      Object {
        "backgroundColor": "white",
        "borderColor": "#e1e8ee",
        "borderWidth": 1,
        "margin": 15,
        "marginBottom": 0,
        "padding": 15,
        "shadowColor": "rgba(0,0,0, .2)",
        "shadowOffset": Object {
          "height": 0,
          "width": 0,
        },
        "shadowOpacity": 1,
        "shadowRadius": 1,
      },
      undefined,
      Array [
        Object {
          "backgroundColor": "#FFFFFF",
          "borderColor": "#D0D1D5",
          "borderRadius": 2,
        },
        Array [],
      ],
    ]
  }>
  <View
    style={
      Array [
        Object {
          "backgroundColor": "transparent",
        },
        undefined,
        undefined,
      ]
    } />
</View>
`;
